---
title: "Burden-function plots"
output:
  html_document:
    df_print: paged
---


```{r}
library(ggplot2)
library(tidyverse)
library(precrec)
library(gridExtra)
library(svglite)
library(rstatix)

source("analysis/star_alleles_burden/utils.R")
```


# OUR5 score
Quantile normalization among all variants per gene
```{r read_data}
pharmvar_g <-
  convert_pharmvar_scores2("data/pharmacogenetic-score/pharmvar.tsv.bgz") %>%
  filter(is_sub) %>%
  select(-AF, -is_sub)
healthy_full <-
  read_tsv(
    "data/pharmacogenetic-score/healthy-non-ref.tsv.bgz",
    # "data/pharmacogenetic-score/healthy-non-ref-filtered.tsv.bgz",
    col_types = "cccdcci"
  ) %>%
  mutate(
    `function` = "healthy",
  )

healthy_genes <-
  healthy_full %>%
  distinct(within_gene) %>%
  pull(within_gene) %>%
  str_split("[^A-Z0-9-\\.]+", simplify = TRUE) %>%
  as.vector() %>%
  unique()
pharmvar_genes <- unique(pharmvar_g$gene_name)
pharmvar <- select(pharmvar_g, -gene_name)
healthy_full <- select(healthy_full, -within_gene)
.all_healthy_subjects <- unique(healthy_full$s)
```

```{r make_score}
dist_to_exon_threshold <- 50
dist_to_gene_threshold <- Inf
af_threshold <- 1.01
pop_af_threshold <- 1.01

genes_list <- list.files("data/pharmacogenetic-score/gene-scores/")
gene_names <- c(
  pharmvar_genes #,
  # "GAPDH", "FKBP5", "VKORC1", "ZBTB16", "DRD1", "SGK1", "TSC22D3", "PLIN4",
  # "HLA-A", "HLA-B", "HLA-DRB1",
  # sample(str_match(genes_list, "(.*).tsv.bgz")[, 2], 10)
)
gene_names_outside <- gene_names[!(gene_names %in% healthy_genes)]
cat(c("Genes not present in healthy: ", paste(gene_names_outside, collapse = ", ")))
gene_names <- unique(
  c(
    gene_names[(gene_names %in% healthy_genes)],
    pharmvar_genes
  )
)

all_scores <- make_score(
  gene_names, dist_exon_th = dist_to_exon_threshold, dist_gene_th = dist_to_gene_threshold
)
```

```{r annotate_data}
healthy_annotated <-
  inner_join(
    healthy_full,
    all_scores,
    by = c("locus", "alleles")
  ) %>%
  unite("star_allele", s, gene_name, remove = FALSE)

pharmvar_annotated <-
  inner_join(
    pharmvar,
    all_scores,
    by = c("locus", "alleles")
  )

df_annotated <-
  bind_rows(pharmvar_annotated, healthy_annotated) %>%
  mutate(pgs = apply(.[BURDEN_SCORE_NAMES], 1, mean, na.rm = TRUE))

pharmvar_annotated %>%
  select(locus,alleles, star_allele, `function`, gene_name) %>%
  filter(`function` != "other") %>%
  write_tsv("results/supplementary/star-alleles-pharmvar.tsv", escape="none")
```

```{r compute_burden}
df_filtered <-
  df_annotated %>%
  filter(is.na(pop_af) | pop_af < pop_af_threshold) %>%
  filter(is.na(gnomad_nfe_AF) | gnomad_nfe_AF < af_threshold)

all_distinct_staralleles <-
  tibble(
    s = rep(.all_healthy_subjects, times = length(gene_names)),
    gene_name = rep(gene_names, each = length(.all_healthy_subjects)),
    `function` = "healthy",
    star_allele = paste(s, gene_name, sep = "_")
  ) %>%
  select(-s) %>%
  bind_rows(pharmvar_g) %>%
  select(star_allele, `function`, gene_name) %>%
  distinct()

our_scores <-
  all_distinct_staralleles %>%  # it's needed because we used *inner_join* earlier

  # left_join(
  #   compute_burden(df_filtered, score_names = BURDEN_SCORE_NAMES),
  #   by = c("star_allele", "function", "gene_name")
  # ) %>%
  mutate(burden = -999) %>%
  left_join(
    compute_burden(df_filtered, score_names = "cadd_raw_oryg"),
    by = c("star_allele", "function", "gene_name"),
    suffix = c("", "_cadd")
  ) %>%
  left_join(
    compute_burden(df_filtered, score_names = "mutation_assessor_oryg"),
    by = c("star_allele", "function", "gene_name"),
    suffix = c("", "_ma")
  ) %>%
  left_join(
    compute_burden(df_filtered, score_names = "provean_oryg"),
    by = c("star_allele", "function", "gene_name"),
    suffix = c("", "_provean")
  ) %>%
  left_join(
    compute_burden(df_filtered, score_names = "fathmm_xf_oryg"),
    by = c("star_allele", "function", "gene_name"),
    suffix = c("", "_fx")
  ) %>%
  left_join(
    compute_burden(df_filtered, score_names = "pgs"),
    by = c("star_allele", "function", "gene_name"),
    suffix = c("", "_pgs")
  ) %>%

  mutate(
    # burden = replace_na(burden, 0),
    burden_pgs = replace_na(burden_pgs, 0),
    burden_cadd = replace_na(burden_cadd, min(burden_cadd, na.rm = TRUE)),
    burden_ma = replace_na(burden_ma, min(burden_ma, na.rm = TRUE)),
    burden_provean = replace_na(burden_provean, min(burden_provean, na.rm = TRUE)),
    burden_fx = replace_na(burden_fx, min(burden_fx, na.rm = TRUE))
  ) %>%
  mutate(across(contains("n_variants"), ~ replace_na(.x, 0)))

our_scores %>%
  filter(`function` != "healthy") %>%
  select(star_allele, `function`, gene_name, burden_pgs) %>%
  filter(`function` != "other") %>%
  filter(gene_name %in% c('CYP2B6', 'CYP2C19', 'CYP2C9', 'CYP2D6', 'CYP3A5')) %>%
  write_csv("results/fig1-values.csv")
```

```{r}
run_info <- paste0(
  c(af_threshold, dist_to_exon_threshold, dist_to_gene_threshold),
  collapse = "_"
)  # put it in a file name
param_str <- paste0(
  "AF: ", af_threshold,
  ", dist to exon: ", dist_to_exon_threshold,
  ", dist to gene: ", dist_to_gene_threshold
)

plot_data <-
  our_scores

ggplot(plot_data, aes(y = burden_pgs, x = `function`)) +
  geom_violin(aes(colour = `function`), fill = 'grey90', size = 1) +
  geom_point(
    aes(size = n_variants),
    position = position_jitter(0.2, seed = 10),
    alpha = 0.5,
    stroke = 0,
  ) +
  theme_minimal() +
  labs(
    title = "Burden (PGS) of star SUB-alleles",
    subtitle = param_str,
    y = "less damaging ←     burden     → more damaging"
  )
ggsave(
  paste0(
    "results/task0/08/", "OUR5-",
    format(Sys.time(), "%d"),
    "-", run_info, "-violins-",
    format(Sys.time(), "%H-%M-%S"),
    ".png"
  ),
  dpi = 180,
  scale = 2,
  height = 10,
  width = 15,
  units = "cm",
  bg = "white"
)

# draw violin plots by genes
plot_function_levels <- c(
  "no function", "decreased function", "healthy",
  "normal function", "increased function", "other"
)
plot_data <-
  our_scores %>%
  filter(`function` != "other") %>%
  mutate(db = if_else(`function` == "healthy", "sportwgs", "pharmvar")) %>%
  mutate(`function` = factor(`function`, levels = plot_function_levels)) %>%
  filter(`function` != "healthy" | n_variants > 0)

ggplot(plot_data, aes(y = burden_pgs, x = gene_name)) +
  geom_point(
    aes(colour = `function`, shape = db),
    position = position_jitter(0.4, seed = 10),
    alpha = 0.8,
    size = 2.0,
    stroke = 1,
  ) +
  scale_shape_manual(values = c(16, 3)) +
  scale_color_manual(values = c("#E41A1C", "#984EA3", "#377EB8", "#4DAF4A", "#FF7F00", "#AAAAAA")) +
  theme_minimal() +
  theme(
    plot.caption.position = "plot"
  ) +
  labs(
    title = "Burden (PGS) of star SUB-alleles (without function 'other')",
    subtitle = param_str,
    y = "less damaging ←     burden     → more damaging",
    caption = date()
  )
ggsave(
  paste0(
    "results/task0/08/", "OUR5-",
    format(Sys.time(), "%d"),
    "-", run_info, "-star-alleles-",
    format(Sys.time(), "%H-%M-%S"),
    ".png"
  ),
  dpi = 180,
  scale = 2,
  height = 10,
  width = 15,
  units = "cm",
  bg = "white"
)
```

```{r manuscript_fig1}
run_info <- paste0(
  c(af_threshold, dist_to_exon_threshold, dist_to_gene_threshold),
  collapse = "_"
)
param_str <- paste0(
  "AF: ", af_threshold,
  ", dist to exon: ", dist_to_exon_threshold,
  ", dist to gene: ", dist_to_gene_threshold
)
# draw violin plots by genes
plot_function_levels <- c(
  "no function", "decreased function",
  "normal function", "increased function", "other"
)
plot_data <-
  our_scores %>%
  filter(`function` != "other") %>%
  filter(`function` != "healthy") %>%
  mutate(`function` = factor(`function`, levels = plot_function_levels))

plot_data %>%
  mutate(cat = paste0(gene_name, `function`)) -> plot_data

ggplot(plot_data %>% group_by(gene_name), aes(y = burden_pgs, x = gene_name)) +
  geom_hline(aes(yintercept = 100), linetype = 2, colour = 'grey20') +
  geom_hline(aes(yintercept = 50), linetype = 'dotted', colour = 'gray20') +
  geom_boxplot(
    aes(colour = `function`),
    outlier.shape = NA,
    show.legend = FALSE
  ) +
  geom_point(
    aes(colour = `function`),
    # position = position_jitter(0.4, seed = 10),
    # position = position_dodge2(padding = 0.3),
    position = position_jitterdodge(),
    alpha = 1,
    size = 1.0,
    shape = 16
  ) +
  scale_shape_manual(values = c(16, 3)) +
  scale_color_manual(values = c("#E41A1C", "#984EA3", "#4DAF4A", "#FF7F00", "#AAAAAA")) +
  theme_minimal() +
  theme(
    plot.caption.position = "plot"
  ) +
  labs(
    title = "Burden (PGS) of star SUB-alleles (without function 'other')",
    subtitle = param_str,
    y = "less damaging ←     burden     → more damaging",
    caption = date()
  )
```

```{r}
ggsave(
  paste0(
    "results/manuscript/", "fig1-",
    format(Sys.time(), "%d"),
    "-", run_info, "-star-alleles-",
    format(Sys.time(), "%H-%M-%S"),
    ".svg"
  ),
  dpi = 180,
  scale = 2,
  height = 7,
  width = 17,
  units = "cm",
  bg = "white"
)
```

```{r manuscript_fig1_statistics}
fig1 <-
  read_csv('results/fig1-values.csv') %>%
  filter(`function` != 'increased function') %>%
  rename(funct = `function`) %>%
  group_by(gene_name)

anova_test(
  fig1,
  dv = burden_pgs,
  wid = star_allele,
  between = funct,
) %>%
write_tsv('results/manuscript/fig1-anova.tsv')

pairwise_t_test(
  fig1 %>% filter(gene_name != 'CYP3A5'),  # only two groups
  burden_pgs ~ funct,
  paired = FALSE,
  pool.sd = FALSE,
  p.adjust.method = 'bonferroni',
  comparisons = list(
    c('no function', 'decreased function'),
    c('no function', 'normal function'),
    c('decreased function', 'normal function')
  )
) %>%
select(-.y.) %>%
write_tsv('results/manuscript/fig1-pairwise-t-test.tsv')
```

# Model evaluation
```{r}
scores_data <-
  our_scores %>%
  mutate(
    `function` = factor(
      `function`,
      levels = c(
        "increased function", "normal function", "healthy",
        "decreased function", "no function", "other"
      )
    )
  )
levels(scores_data$`function`) <- c(
        "normal function", "normal function", "healthy",
        "no function", "no function", "other"
)

scores_data <-
  scores_data %>%
  filter(`function` != "healthy") %>%
  filter(`function` != "other")

# in case of precrec's mmdata function error update Rcpp package
# install.packages("Rcpp")
precrec_obj_ <- mmdata(
  scores = list(
    scores_data$burden_cadd,
    scores_data$burden_fx,
    scores_data$burden_provean,
    scores_data$burden_ma,
    scores_data$burden_pgs
    # scores_data$burden
  ),
  labels = scores_data$`function`,
  modnames = c('cadd', 'fathmm_xf', 'provean', 'mutation_assesor', 'PGS')
)
precrec_obj <- evalmod(precrec_obj_)
data.frame(evalmod(precrec_obj_, mode = "aucroc"))
p_roc <- autoplot(precrec_obj, ret_grob = TRUE, curvetype = "ROC")
p_prc <- autoplot(precrec_obj, ret_grob = TRUE, curvetype = "PRC")
p <- grid.arrange(
  p_roc + labs(subtitle = param_str) + theme(legend.position = "bottom"),
  p_prc + labs(subtitle = param_str) + theme(legend.position = "bottom"),
  nrow = 1
)

ggsave(
  paste0(
    "results/manuscript/", "fig1-auc-",
    format(Sys.time(), "%d"),
    "-", run_info, "-eval-",
    format(Sys.time(), "%H-%M-%S"),
    ".svg"
  ),
  plot = p,
  width = 20,
  height = 12,
  units = "cm",
  dpi = 600,
  bg = "white"
)
```

# Normalize scores for all genes
```{r}
genes_to_score <- sub("\\.tsv\\.bgz$", "", dir("data/pharmacogenetic-score/gene-scores/"))
# genes_to_score <- genes_to_score[16580:length(genes_to_score)]
for (gene in genes_to_score) {
  if (str_detect(gene, "TEST_GENE")){
    next
  }
  make_score(
    gene,
    dist_exon_th = dist_to_exon_threshold,
    dist_gene_th = dist_to_gene_threshold
  ) %>%
  separate(alleles, c("X1", "allele_1", "allele_2", "X4")) %>%
  select(-X1, -X4) %>%
  write_tsv(
    paste0("data/pharmacogenetic-score/gene-our-scores/", gene, ".tsv"),
    escape = "none"
  )
}
```
